ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionBroadcaster.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright contributors to Hyperledger Besu
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.eth.transactions;
+
+import org.hyperledger.besu.ethereum.core.Transaction;
+import org.hyperledger.besu.ethereum.eth.manager.EthContext;
+import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
+import org.hyperledger.besu.ethereum.eth.messages.EthPV65;
+import org.hyperledger.besu.ethereum.eth.transactions.TransactionPool.TransactionBatchAddedListener;
+import org.hyperledger.besu.ethereum.eth.transactions.sorter.AbstractPendingTransactionsSorter;
+import org.hyperledger.besu.ethereum.eth.transactions.sorter.AbstractPendingTransactionsSorter.TransactionInfo;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class TransactionBroadcaster implements TransactionBatchAddedListener {
+
+  private final AbstractPendingTransactionsSorter pendingTransactions;
+  private final PeerTransactionTracker transactionTracker;
+  private final TransactionsMessageSender transactionsMessageSender;
+  private final NewPooledTransactionHashesMessageSender newPooledTransactionHashesMessageSender;
+  private final EthContext ethContext;
+  private final int numPeersToSendFullTransactions;
+
+  public TransactionBroadcaster(
+      final EthContext ethContext,
+      final AbstractPendingTransactionsSorter pendingTransactions,
+      final PeerTransactionTracker transactionTracker,
+      final TransactionsMessageSender transactionsMessageSender,
+      final NewPooledTransactionHashesMessageSender newPooledTransactionHashesMessageSender) {
+    this.pendingTransactions = pendingTransactions;
+    this.transactionTracker = transactionTracker;
+    this.transactionsMessageSender = transactionsMessageSender;
+    this.newPooledTransactionHashesMessageSender = newPooledTransactionHashesMessageSender;
+    this.ethContext = ethContext;
+    this.numPeersToSendFullTransactions =
+        (int) Math.ceil(Math.sqrt(ethContext.getEthPeers().getMaxPeers()));
```

Comment:
I was just thinking that in case I would like to manually set up this value to some concrete number, I cannot because its calculation is hardcoded in the constructor. I don't think if I ever should or if it makes sense to do it.

Reply:
the value is taken from https://eips.ethereum.org/EIPS/eip-2464 where it suggests to send full transactions only to the square root of peers. Not sure we want to ever change this value, but in that case we could make it configurable in future.